sentinel/src/lib/history.js

```diff
@@ -0,0 +1,66 @@
+/**
+ * Utility to track duplicate keys in a certain period of time (periodMins).
+ * Removes expired keys after threshold is reached (purgeThreshold).
+ **/
+const moment = require('moment');
+
+let records = {};
+
+const now = () => {
+  return moment(new Date());
+};
+
+const expired = (key) => {
+  const value = records[key];
+  if(value && value.timestamp) {
+    // Checks difference with history duration in minutes
+    return (now().diff(value.timestamp))/60000 >= module.exports.periodMins;
+  }
+  return true;
+};
+
+const size = () => {
+  return Object.keys(records).length;
+};
+
+const purge = (force=false) => {
+  if(force || size() >= module.exports.purgeThreshold) {
+    for (const key of Object.keys(records)) {
+      if(expired(key)) {
+        delete records[key];
+      }
+    }
+    return true;
+  }
+  return false;
+};
+
+module.exports = {
```

Comment:
Quite a few of the exported functions are only used for testing. We have a convention of prefixing these with underscores (so `_periodMins: 30`). Also they should be ordered to the bottom of the exports with a note above that they are `// exposed for testing` or similar.